G-API: Build warnings with GCC 7.5 (Ubuntu 18.04)

[Nigthly build](http://pullrequest.opencv.org/buildbot/builders/master_opt-avx2_noICV-skl-lin64/builds/776):

```
[1186/1744] Building CXX object modules/gapi/CMakeFiles/opencv_test_gapi.dir/test/cpu/gapi_ocv_stateful_kernel_tests.cpp.o
In file included from /build/master_opt-avx2_noICV-skl-lin64/opencv/modules/gapi/include/opencv2/gapi/gcommon.hpp:17:0,
                 from /build/master_opt-avx2_noICV-skl-lin64/opencv/modules/gapi/include/opencv2/gapi/gmat.hpp:15,
                 from /build/master_opt-avx2_noICV-skl-lin64/opencv/modules/gapi/include/opencv2/gapi.hpp:25,
                 from /build/master_opt-avx2_noICV-skl-lin64/opencv/modules/gapi/test/cpu/../test_precomp.hpp:18,
                 from /build/master_opt-avx2_noICV-skl-lin64/opencv/modules/gapi/test/cpu/gapi_ocv_stateful_kernel_test_utils.hpp:10,
                 from /build/master_opt-avx2_noICV-skl-lin64/opencv/modules/gapi/test/cpu/gapi_ocv_stateful_kernel_tests.cpp:7:
/build/master_opt-avx2_noICV-skl-lin64/opencv/modules/gapi/include/opencv2/gapi/util/any.hpp:70:21: warning: 'cv::util::any::holder_ptr cv::util::any::holder_impl<value_t>::clone() [with value_t = opencv_test::{anonymous}::UserStruct]' declared 'static' but never defined [-Wunused-function]
          holder_ptr clone() override { return holder_ptr(new holder_impl (v));}
                     ^~~~~
```

---

[GCC 8 is affected too](http://pullrequest.opencv.org/buildbot/builders/master_etc-fedora-tbb-lin64/builds/779).

Try GCC 7.x and force AVX2 like nightly builder does.
Ubuntu 18.04 docker env:

```
$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu Bionic Beaver (development branch)"
$ which gcc
/usr/bin/gcc
$ gcc --version
gcc (Ubuntu 7.3.0-16ubuntu3) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ cmake -GNinja ../../dev -DCMAKE_BUILD_TYPE=Release -DCPU_BASELINE=AVX2 -DCPU_DISPATCH=OFF -DWITH_IPP=OFF
...

$ ninja -v modules/gapi/CMakeFiles/opencv_test_gapi.dir/test/cpu/gapi_ocv_stateful_kernel_tests.cpp.o
[1/1] /usr/bin/ccache /usr/bin/c++  -D_USE_MATH_DEFINES -D__OPENCV_BUILD=1 -D__OPENCV_TESTS=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -isystem . -isystem /usr/include/x86_64-linux-gnu -I/home/alalek/projects/opencv/dev/modules/ts/include -I/home/alalek/projects/opencv/dev/modules/gapi/include -I/home/alalek/projects/opencv/dev/modules/imgcodecs/include -I/home/alalek/projects/opencv/dev/modules/videoio/include -I/home/alalek/projects/opencv/dev/modules/core/include -I/home/alalek/projects/opencv/dev/modules/flann/include -I/home/alalek/projects/opencv/dev/modules/imgproc/include -I/home/alalek/projects/opencv/dev/modules/dnn/include -I/home/alalek/projects/opencv/dev/modules/features2d/include -I/home/alalek/projects/opencv/dev/modules/calib3d/include -I/home/alalek/projects/opencv/dev/modules/video/include -I/home/alalek/projects/opencv/dev/modules/highgui/include -I/home/alalek/projects/opencv/dev/modules/gapi/src -isystem /usr/include/eigen3 -I3rdparty/ade/ade-0.1.1f/sources/ade/include -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Winit-self -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mf16c -mfma -mavx -mavx2 -fvisibility=hidden -fvisibility-inlines-hidden -O3 -DNDEBUG  -DNDEBUG -fPIE   -std=c++11 -MD -MT modules/gapi/CMakeFiles/opencv_test_gapi.dir/test/cpu/gapi_ocv_stateful_kernel_tests.cpp.o -MF modules/gapi/CMakeFiles/opencv_test_gapi.dir/test/cpu/gapi_ocv_stateful_kernel_tests.cpp.o.d -o modules/gapi/CMakeFiles/opencv_test_gapi.dir/test/cpu/gapi_ocv_stateful_kernel_tests.cpp.o -c /home/alalek/projects/opencv/dev/modules/gapi/test/cpu/gapi_ocv_stateful_kernel_tests.cpp
In file included from /home/alalek/projects/opencv/dev/modules/gapi/include/opencv2/gapi/gcommon.hpp:17:0,
                 from /home/alalek/projects/opencv/dev/modules/gapi/include/opencv2/gapi/gmat.hpp:15,
                 from /home/alalek/projects/opencv/dev/modules/gapi/include/opencv2/gapi.hpp:25,
                 from /home/alalek/projects/opencv/dev/modules/gapi/test/cpu/../test_precomp.hpp:18,
                 from /home/alalek/projects/opencv/dev/modules/gapi/test/cpu/gapi_ocv_stateful_kernel_test_utils.hpp:10,
                 from /home/alalek/projects/opencv/dev/modules/gapi/test/cpu/gapi_ocv_stateful_kernel_tests.cpp:7:
/home/alalek/projects/opencv/dev/modules/gapi/include/opencv2/gapi/util/any.hpp:70:21: warning: 'cv::util::any::holder_ptr cv::util::any::holder_impl<value_t>::clone() [with value_t = opencv_test::{anonymous}::UserStruct]' declared 'static' but never defined [-Wunused-function]
          holder_ptr clone() override { return holder_ptr(new holder_impl (v));}
                     ^~~~~
```

--- a/modules/gapi/test/cpu/gapi_ocv_stateful_kernel_test_utils.hpp
+++ b/modules/gapi/test/cpu/gapi_ocv_stateful_kernel_test_utils.hpp
@@ -22,8 +22,10 @@ struct Name                                      \
 
 namespace opencv_test
 {
-namespace
-{
+
+// types from anonymous namespace doesn't work well with templates
+inline namespace gapi_ocv_stateful_kernel_test_utils {
+
 struct UserStruct
 {
     UserStruct() = default;
@@ -41,7 +43,8 @@ private:
     short _myShortVal;
     float _myFloatVal;
 };
-} // anonymous namespace
+
+} // namespace
 } // opencv_test
 
 #endif // OPENCV_GAPI_OCV_STATEFUL_KERNEL_TESTS_UTILS_HPP

--- a/modules/gapi/test/gapi_array_tests.cpp
+++ b/modules/gapi/test/gapi_array_tests.cpp
@@ -240,7 +240,8 @@ TEST(GArray_VectorRef, TestMov)
     EXPECT_EQ(V{}, vtest);
 }
 
-namespace {
+// types from anonymous namespace doesn't work well with templates
+inline namespace gapi_array_tests {
     struct MyTestStruct {
         int i;
         float f;

--- a/modules/gapi/test/gapi_opaque_tests.cpp
+++ b/modules/gapi/test/gapi_opaque_tests.cpp
@@ -284,7 +284,8 @@ TEST(GOpaque_OpaqueRef, TestMov)
     EXPECT_NE(test, mov.rref<I>());         // ref lost the data
 }
 
-namespace {
+// types from anonymous namespace doesn't work well with templates
+inline namespace gapi_opaque_tests {
     struct MyTestStruct {
         int i;
         float f;

